### PR TITLE
[ALLUXIO-2997] [s3] Implement S3 DELETE object API

### PR DESCRIPTION
**Pull request by ifcharming**
https://alluxio.atlassian.net/browse/ALLUXIO-2997

A valid object is either an existing Alluxio file or an empty Alluxio directory.
Also added tests.

How this PR is tested:
Along with the other [PR](https://github.com/Alluxio/alluxio/pull/5927) is merged, the following script works (assuming `/bucket1/key1` is a file exists in Alluxio) :
```import boto
import boto.s3.connection
access_key = 'put your access key here!'
secret_key = 'put your secret key here!'

conn = boto.connect_s3(
    aws_access_key_id = access_key,
    aws_secret_access_key = secret_key,
    host = 'localhost',
    port = 39999,
    path = '/api/v1/s3',
    is_secure=False,
    calling_format = boto.s3.connection.OrdinaryCallingFormat(),
)

# This will succeed.
bucket = conn.get_bucket('bucket1')

print bucket
for key in bucket.list():
    print "{name}\t{size}\t{modified}".format(
        name = key.name,
        size = key.size,
        modified = key.last_modified,
        )

bucket.delete_key('key1')
```

### CODE REVIEW

**Comment by ifcharming** [comment]
alluxio-bot, test this please

**Comment by ifcharming** [comment]
Jenkins, test this please

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 1]
```diff
@@ -178,4 +232,23 @@ private void checkBucketIsAlluxioDirectory(String bucketPath) throws S3Exception
...
     }
   }
+
+  /**
+   * Checks the given object is either a file or an empty directory in Alluxio.
+   *
+   * @param objectPath the object path in Alluxio
+   * @throws S3Exception if the object is not a valid in Alluxio
+   */
+  private void checkObjectInAlluxio(String objectPath) throws S3Exception {
```
Comment (calvinjia) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: What guarantee do we get by having this method? Invocations to this method do not guarantee future invocations with the same path exist or do not exist.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 2]
```diff
@@ -79,11 +81,11 @@ public S3RestServiceHandler(@Context ServletContext context) {
...
-  @ReturnType("java.lang.Void")
+  @ReturnType("Response.Status")
   public Response createBucket(@PathParam("bucket") final String bucket) {
-    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Void>() {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response.Status>() {
       @Override
-      public Void call() throws S3Exception {
+      public Response.Status call() throws S3Exception {
         String bucketPath =  AlluxioURI.SEPARATOR + bucket;
         if (bucketPath.contains(BUCKET_SEPARATOR)) {
```
Comment (madanadit) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: create a utility function for this logic?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 1 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: True, it's possible that this method passed but `delete` throws `FileDoesNotExistException`.
But this is acting more like a bloom filter to filter out the "non-empty directory" cases.  We don't want to view that as a valid object and try deleting it.

Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: indeed this method introduces 1-2 more Alluxio RPCs, I'm open to removing this and reply on the `delete` to detect `FileDoesNotExistException` and `DirectoryNotEmptyException`. Do you have a preference?

Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: Removed for now... now deleting a non-empty directory will result in `FAILED_PRECONDITION`, not a invalid key. I think it should be okay.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 2 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: done

**Comment by ifcharming** [comment]
@calvinjia @madanadit I've addressed your comments, can you PTAL?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 3]
```diff
@@ -156,6 +182,29 @@ private S3Exception toBucketS3Exception(Exception exception, String resource) {
...
+    } catch (DirectoryNotEmptyException e) {
+      return new S3Exception(resource, S3ErrorCode.PRECONDITION_FAILED);
+    } catch (FileDoesNotExistException e) {
+      return new S3Exception(resource, S3ErrorCode.NO_SUCH_KEY);
+    } catch (Exception e) {
+      return new S3Exception(e, resource, S3ErrorCode.INTERNAL_ERROR);
+    }
+  }
+
+  private String maybeUpdateAndCheckBucketPath(String bucketPath) throws S3Exception {
```
Comment (madanadit) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: nit: maybe rename to `parseBucketPath`. javadoc?

Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: done

**Comment by ifcharming** [comment]
@calvinjia can you please take another look? thanks!

**Comment by ifcharming** [comment]
Merging this PR to unblock other object operations... @calvinjia please do a post-merge review. thanks

**Pull request merged**